### ISSUE TITLE
Request Validators

### ISSUE THREAD

**Issue by t1** [eclipse/microprofile-graphql#236]
Add a mechanism to generically inspect requests in order to, e.g. reject queries that can be too expensive to execute by limiting them to some maximum query depth (see [here](https://www.howtographql.com/advanced/4-security/)). Maybe something like this:

```java
class MaximumQueryDepthValidator implements Validator {
    @Inject @ConfigPropert Integer maxQueryDepth;

    @Override public void validate(GraphQlRequest request) {
        if (request.getDepth() > maxQueryDepth) {
            throw new RuntimeException("maximum query depth exceeded");
        }
    }
}
```

I assume that this would need to standardize and expose a lot of data structures that are yet internal implementation details: the validators would need to have access to the request, the schema, and the methods or at least the annotations on the methods. E.g. to compute the total complexity of a request, a validator could sum the values of a custom `@Complexity` annotation on all `@Source` methods that will be called.

In a later step, we could also add some standard validators.

**Comment by ybroeker** [comment]
It might be useful to split that in two?

A few config options that would cover most basic cases and work "out of the box", perhaps maxQueryDepth and maxFields (which would be the same as maxComplexity with complexity 1 for each field), and validation separately.

The former would be a relatively lightweight change, both for the spec and for the implementations.  Validation would be a bigger change, and would probably have to build on #127 as well?

**Comment by t1** [comment]
You're right. Probably it makes sense to start with specific use-cases and generalize from there.
